### PR TITLE
XML Files

### PR DESCRIPTION
**Pull request by EliLentz**
1. Added the ability to generate bunnies in advance, using XML
2. Removed the ability to generate bunnies using the "GenerateInitialBunnies" function
3. Now the life cycle and the change of years pass automatically (every 3 seconds)
4. Fixed bug with bunny breeding. (the number of new individuals depended solely on the number of males)

### CODE REVIEW

File: Hill.cs [hunk 1]
```diff
@@ -34,10 +35,14 @@ public Hill(List<Bunny> bunnies)
         /// </summary>
         private void TheCycleOfLife()
```
Comment (igalvino) [inline Hill.cs]: RunCycle()*

File: Hill.cs [hunk 2]
```diff
@@ -34,10 +35,14 @@ public Hill(List<Bunny> bunnies)
         /// </summary>
         private void TheCycleOfLife()
         {
-            while (_hill.Count != 0)//circle in endless rage
+            while (_hill.Count != 0)
```
Comment (igalvino) [inline Hill.cs]: while(true)

File: IntegeratorXml.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
+﻿using Bunnies;
+using bunnys_hill;
+using System;
+using System.Collections.Generic;
+using System.Xml;
+
+namespace Integerator
+{
+    class IntegeratorXml
```
Comment (igalvino) [inline IntegeratorXml.cs]: *XMLReader

File: IntegeratorXml.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        /// <returns></returns>
+        private static int[] SearcherLong(XmlNode xnode)
+        {
+            int[] attrs = new int[3];
+
+            foreach (XmlNode childNode in xnode.ChildNodes)
+            {
+                if (childNode.Name == "Sex")
+                {
+                    attrs[0] = (int)Enum.Parse(typeof(Sex), childNode.InnerText);
```
Comment (igalvino) [inline IntegeratorXml.cs]: Use TryParse instead

File: IntegeratorXml.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            xDoc.Load(link);
+
+            XmlElement xRoot = xDoc.DocumentElement;
+
+            foreach (XmlNode xnode in xRoot)
+            {
+                XmlNode mainAttr = xnode.Attributes.GetNamedItem("Name");
+                selectedName = mainAttr.Value;
+
+                dataBunny = SearcherLong(xnode);
```
Comment (igalvino) [inline IntegeratorXml.cs]: can be removed

File: IntegeratorXml.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        #endregion
+
+        #region HelpFunctions
+
+        /// <summary>
+        /// this function searching attributs if they are written in long form
+        /// </summary>
+        /// <param name="xnode"></param>
+        /// <returns></returns>
+        private static int[] SearcherLong(XmlNode xnode)
```
Comment (igalvino) [inline IntegeratorXml.cs]: can be removed

File: IntegeratorXml.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+            XmlElement xRoot = xDoc.DocumentElement;
+
+            foreach (XmlNode xnode in xRoot)
+            {
+                XmlNode mainAttr = xnode.Attributes.GetNamedItem("Name");
+                selectedName = mainAttr.Value;
+
+                dataBunny = SearcherLong(xnode);
+                dataBunny = SearcherShort(xnode);
```
Comment (igalvino) [inline IntegeratorXml.cs]: Bunny ParseSingleBunny(xnode)

File: IntegeratorXml.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+                Bunny selectedBunny = new Bunny(selectedName, dataBunny[0], dataBunny[1], dataBunny[2]);
+                initialBunnies.Add(selectedBunny);
+            }
+
+            Logic.PrintNewBunnies(initialBunnies);
```
Comment (igalvino) [inline IntegeratorXml.cs]: Dont call the print from the reader.
Reader should only do read.

File: IntegeratorXml.cs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        /// <summary>
+        /// this function searching attributs if they are written in short form
+        /// </summary>
+        /// <param name="xnode"></param>
+        /// <returns></returns>
+        private static int[] SearcherShort(XmlNode xnode)
+        {
+            int[] attrs = new int[3];
```
Comment (igalvino) [inline IntegeratorXml.cs]: int age = 0;
string name....


name = Int32.TryParse(childAttr.InnerText);


return new Bunny(name...);

File: Logic.cs [hunk 10]
```diff
@@ -10,6 +12,8 @@ public class Logic
         public static int countNewBunnies = 0;//how many bunnies were born
         public static int countDeadBunnies = 0;//how many bunnies were died
 
+        private static volatile bool stopFlow = false;//Regulates flow stopping
```
Comment (igalvino) [inline Logic.cs]: try without volatile, and return with answers.

File: Logic.cs [hunk 11]
```diff
@@ -123,6 +90,46 @@ public static void KillOldBunnies(List<Bunny> currentBunnies)
...
+        /// <summary>
+        /// This function pass a year every 3 seconds
+        /// </summary>
+        public static void TimesHill()
+        {
+            Thread.Sleep(3000);
+
+            while (stopFlow)
+            {
+                Task taskStopper = Stopper();
```
Comment (igalvino) [inline Logic.cs]: This should only happen once, not in a loop.

File: Program_1.cs [hunk 12]
```diff
@@ -9,7 +10,11 @@ static void Main(string[] args)
         {
             List<Bunny> bunnies;//creating initial bunnies
 
-            bunnies = Logic.GenerateInitialBunnies(5); //giving random features (color and gender)for the bunnies
+            //bunnies = Logic.GenerateInitialBunnies(5); //giving random features (color and gender)for the bunnies
+
+            string xmlFile = "InitailBunnies.xml"; //file's URL address
```
Comment (igalvino) [inline Program_1.cs]: should be const string FILE_NAME = ...

File: Logic.cs [hunk 13]
```diff
@@ -123,6 +90,46 @@ public static void KillOldBunnies(List<Bunny> currentBunnies)
...
         #endregion
 
+        #region TimeOfHill
+
+        /// <summary>
+        /// This function pass a year every 3 seconds
+        /// </summary>
+        public static void TimesHill()
+        {
+            Thread.Sleep(3000);
```
Comment (igalvino) [inline Logic.cs]: 3000 should be const

File: Logic.cs [hunk 14]
```diff
@@ -123,6 +90,46 @@ public static void KillOldBunnies(List<Bunny> currentBunnies)
         }
         #endregion
 
+        #region TimeOfHill
+
+        /// <summary>
+        /// This function pass a year every 3 seconds
+        /// </summary>
+        public static void TimesHill()
```
Comment (igalvino) [inline Logic.cs]: *find better name

**Review by igalvino — COMMENTED** [review]
The comments

File: Logic.cs (hunk 10 above)
Comment (EliLentz) [reply Logic.cs]: It's working without "volatile"

**Pull request merged**